ROCprofiler-SDK: deprecation of legacy tools (#3609)
ROCm/rocm-systems: projects/rocprofiler-sdk/source/docs/conceptual/comparing-with-legacy-tools.rst

Reviewer: ```suggestion
- Perform more extensive checks during service specification and inform the tool about potential issues early.
```

.. _comparing-with-legacy-tools:

=========================================================
Comparing ROCprofiler-SDK to legacy ROCm profiling tools
=========================================================

> ROCprofiler-SDK is an improved version of ROCm profiling tools that enables more efficient implementations and better thread safety while avoiding problems that plagued the legacy ROCm profiling tools: ROCProfiler and ROCTracer.

Reviewer: ```suggestion
ROCprofiler-SDK is an improved version of the ROCm profiling tools that enables more efficient implementations and better thread safety while avoiding the problems that hindered the legacy ROCm profiling tools: ROCProfiler and ROCTracer.
```

Reviewer: Both of these should be linked to the docs:
ROCProfiler and ROCTracer

- Backward ABI compatibility
- PC sampling (beta implementation)

The legacy ROCm profiling tools allowed a tool to access any of the services provided by ROCProfiler or ROCTracer, such as API tracing and kernel tracing, by calling ``roctracer_init()`` when an ROCm runtime is initially loaded.

> As the calling tool is not required to specify during initialization, the services it needs to use, the libraries must be effectively prepared for any service to be available anytime.

Reviewer: ```suggestion
Because the calling tool is not required to specify the services it needs during initialization, the libraries must be effectively prepared for any service to be available at any time.
```

Author: This changes the meaning. Those services are not needed during initialization. Those are the services it will use later on.

Author: Changed it to: "Since the services to be used by the calling tool are not required to be specified during initialization, the libraries must be effectively prepared for any service to be available at any time."

This behavior introduced unnecessary overhead and made thread-safe data management difficult, as tools generally don't use all the available services.
For example, ROCTracer always installed wrappers around every runtime API and added indirection overhead through the ROCTracer library to check for the current service configuration in a thread-safe manner.

ROCprofiler-SDK introduces ``context`` to solve the preceding issues. Contexts are effectively bundles of service configurations. ROCprofiler-SDK provides a single opportunity for a tool to create as many contexts as required.

> A tool can group all services into one context, create one context per service, or choose a mix.

Reviewer: ```suggestion
A tool can group all services into a single context, create a separate context for each service, or use a mix.
```

- Improve thread safety without introducing parallel bottlenecks.
- Manage internal data and allocations more efficiently.

> Command-line tool options: rocprofv3 vs. rocprof and rocprofv2

Reviewer: Why doesn't the page intro mention this page will compare tools and command line tools?

Reviewer: versus instead of vs.

===============================================================

> ROCprofiler-SDK introduces a new command-line (CLI) tool, ``rocprofv3``, which is more efficient and flexible than the legacy ROCProfiler CLI tools: ``rocprof`` and ``rocprofv2``.

Reviewer: ```suggestion
ROCprofiler-SDK includes a command-line (CLI) tool, ``rocprofv3``, which is more efficient and flexible than the legacy ROCProfiler CLI tools: ``rocprof`` and ``rocprofv2``.
```

> .. list-table:: Comparison of ROCprofiler Command-Line Tool's options

Reviewer: ```suggestion
.. list-table:: Comparing ROCprofiler CLI options
```

   * - Trace control options
     - Merge traces
     - ``--merge-traces``
     - *Not available*
     - *Not available*
     - Not available in ``rocprofv3``
     -
   * - PC sampling options
     - PC sampling
     - *Not available*
     - *Not available*
     - ``--pc-sampling-beta-enabled``
     - Enables PC sampling support; beta version.
     -
   * - Legacy options
     - Timestamp On/Off
     - ``--timestamp <on|off>``
     - *Not available*
     - *Not available*
     - Not applicable for ``rocprofv3``.
     -
   * - Legacy options
     - Context wait
     - ``--ctx-wait``
     - *Not available*
     - *Not available*
     - Not applicable for ``rocprofv3``.
     -
   * - Legacy options
     - Context limit
     - ``--ctx-limit <max number>``
     - *Not available*
     - *Not available*
     - Not applicable for ``rocprofv3``.
     -
   * - Legacy options
     - Code object tracking
     - ``--obj-tracking <on|off>``
     - Always ``ON`` in ``rocprofv2``.
     - Always ``ON`` in ``rocprofv3``.
     -
     -
   * - Legacy options
     - Heartbeat
     - ``--heartbeat <rate sec>``
     - *Not available*
     - *Not available*
     - Not applicable for ``rocprofv3``
     -

> Timing information: rocprofv3 vs. rocprof and rocprofv2

Reviewer: ```suggestion
Timing information: rocprofv3 versus rocprof and rocprofv2
```

========================================================

> ``rocprofv3`` provides more accurate timing information by reducing the tool overhead required to collect data along with the interference in the timing of the kernel being measured. This results in a reduced kernel time variance received for the same kernel execution and more accurate timing in general. These changes are not available in ``rocprof`` and ``rocprofv2``, so there can be substantial differences (upto 20%) in execution times reported by ``rocprof`` and ``rocprofv2`` for a single kernel execution in comparison with ``rocprofv3``. Over a large number of samples of the same kernel, the difference in the average execution time is limited to a single digit with a much tighter variance of results on ``rocprofv3``.

Reviewer: ```suggestion
``rocprofv3`` provides more accurate timing information by reducing the tool overhead required to collect data and the interference with the kernel's timing. This results in a reduced kernel time variance received for the same kernel execution and more accurate timing overall. These changes are not available in ``rocprof`` and ``rocprofv2``, so there can be substantial differences (up to 20 percent) in execution times reported by ``rocprof`` and ``rocprofv2`` for a single kernel execution in comparison with ``rocprofv3``. Across a large number of samples of the same kernel, the difference in average execution time is limited to a single digit, with a much tighter variance of results on ``rocprofv3``.
```

> Default behavior: rocprofv3 vs. rocprof and rocprofv2

Reviewer: ```suggestion
Default behavior: rocprofv3 versus rocprof and rocprofv2
```

======================================================

> When being run without an option, ``rocprofv3`` behaves differently than ``rocprof`` and ``rocprofv2``. The default behavior of ``rocprofv3`` is to collect all available agents on the system and output it in the ``.csv`` format while ``rocprof`` and ``rocprofv2`` output the kernel traces in ``.csv`` format by default. On ``rocprofv3``, kernel traces are generated using ``--kernel-trace`` option.

Reviewer: ```suggestion
When run without an option, ``rocprofv3`` behaves differently than ``rocprof`` and ``rocprofv2``. The default behavior of ``rocprofv3`` is to collect all available agents on the system and output them in ``CSV`` format, while ``rocprof`` and ``rocprofv2`` output the kernel traces in ``CSV`` format by default. On ``rocprofv3``, kernel traces are generated using ``--kernel-trace`` option.
```
